fix new_deadline_info_from_offset_and_epoch

zhiqiangxu opened the pull request:
This PR introduces a new field `wpost_deadline_window` to the `Policy` struct. Previously there was a [bug](https://github.com/filecoin-project/builtin-actors/blob/b712d1ed4ad6f6da57916caaed8af209daaf4db4/actors/miner/src/deadlines.rs#L142) but it happens that `wpost_challenge_window` and `wpost_deadline_window` are equal but they don't have to be.

`% policy.wpost_period_deadlines` is unnecessary since `current_epoch - current_period_start` is in range [0, wpost_proving_period), so `((current_epoch - current_period_start) / policy.wpost_deadline_window)` is in range [0, wpost_period_deadlines)

anorth commented:
They do have to be the same. What you have called the deadline window _is_ the challenge window (maybe it's not a great name). The value `policy.wpost_challenge_window` is used as a divisor in other places in the code too, not just the one you found.

It appears that some critical code was lost in the translation from the Go specs actors. Here the [definition of WPostChallengeWindow](https://github.com/filecoin-project/specs-actors/blob/985cd0fa04578e262d68e0ef196f17df6f2434f2/actors/builtin/miner/policy.go#L19-L22) has better documentation.

> // The period between the opening and the closing of a WindowPoSt deadline in which the miner is expected to
> // provide a Window PoSt proof.
> // This provides a miner enough time to compute and propagate a Window PoSt proof.
> var WPoStChallengeWindow = abi.ChainEpoch(30 * 60 / builtin.EpochDurationSeconds) // 30 minutes (48 per day)

And below you can see runtime assertions that this challenge window evenly divides the proving period exactly `WPostPeriodDeadlines` times.

> // Check that the challenge windows divide the proving period evenly.
> if WPoStProvingPeriod%WPoStChallengeWindow != 0 {
>     panic(fmt.Sprintf("incompatible proving period %d and challenge window %d", WPoStProvingPeriod, WPoStChallengeWindow))
> }
> // Check that WPoStPeriodDeadlines is consistent with the proving period and challenge window.
> if abi.ChainEpoch(WPoStPeriodDeadlines)*WPoStChallengeWindow != WPoStProvingPeriod {
>     panic(fmt.Sprintf("incompatible proving period %d and challenge window %d", WPoStProvingPeriod, WPoStChallengeWindow))
> }

So the real problem here is the loss of documentation and assertions. This appears to be the case at the [initial import](https://github.com/filecoin-project/builtin-actors/commit/3e6b251c9bed4c8e99c234b53ac8046b89050f7e#diff-63cb0bfdfcb9eb44a7f45da302e7cd12be670e9ea74b93419fb6435841c6fab6) of this code into this repository.

If you could fix the documentation right now, that would help avoid future confusion. But we should also re-instate these consistency assertions somehow. @raulk @Stebalien do you have suggestions here?

zhiqiangxu commented:
> They do have to be the same. What you have called the deadline window _is_ the challenge window (maybe it's not a great name). The value `policy.wpost_challenge_window` is used as a divisor in other places in the code too, not just the one you found.
> 
> It appears that some critical code was lost in the translation from the Go specs actors. Here the [definition of WPostChallengeWindow](https://github.com/filecoin-project/specs-actors/blob/985cd0fa04578e262d68e0ef196f17df6f2434f2/actors/builtin/miner/policy.go#L19-L22) has better documentation.
> 
> > // The period between the opening and the closing of a WindowPoSt deadline in which the miner is expected to
> > // provide a Window PoSt proof.
> > // This provides a miner enough time to compute and propagate a Window PoSt proof.
> > var WPoStChallengeWindow = abi.ChainEpoch(30 * 60 / builtin.EpochDurationSeconds) // 30 minutes (48 per day)
> 
> And below you can see runtime assertions that this challenge window evenly divides the proving period exactly `WPostPeriodDeadlines` times.
> 
> > // Check that the challenge windows divide the proving period evenly.
> > if WPoStProvingPeriod%WPoStChallengeWindow != 0 {
> > panic(fmt.Sprintf("incompatible proving period %d and challenge window %d", WPoStProvingPeriod, WPoStChallengeWindow))
> > }
> > // Check that WPoStPeriodDeadlines is consistent with the proving period and challenge window.
> > if abi.ChainEpoch(WPoStPeriodDeadlines)*WPoStChallengeWindow != WPoStProvingPeriod {
> > panic(fmt.Sprintf("incompatible proving period %d and challenge window %d", WPoStProvingPeriod, WPoStChallengeWindow))
> > }
> 
> So the real problem here is the loss of documentation and assertions. This appears to be the case at the [initial import](https://github.com/filecoin-project/builtin-actors/commit/3e6b251c9bed4c8e99c234b53ac8046b89050f7e#diff-63cb0bfdfcb9eb44a7f45da302e7cd12be670e9ea74b93419fb6435841c6fab6) of this code into this repository.
> 
> If you could fix the documentation right now, that would help avoid future confusion. But we should also re-instate these consistency assertions somehow. @raulk @Stebalien do you have suggestions here?

That's true. I reverted the addition of `wpost_deadline_window` and added a comment about the assumption.

`% policy.wpost_period_deadlines` is still unnecessary nevertheless.
